mypy/stubgenc.py

```diff
@@ -182,439 +158,649 @@ def generate_stub_for_c_module(
+                inferred_type = self.add_name("_typeshed.Incomplete")
+
+            static_properties.append(
+                f"{self._indent}{name}: {classvar}[{inferred_type}] = ...{trailing_comment}"
             )
-        )
-    all_bases = type.mro(obj)
-    if all_bases[-1] is object:
-        # TODO: Is this always object?
-        del all_bases[-1]
-    # remove pybind11_object. All classes generated by pybind11 have pybind11_object in their MRO,
-    # which only overrides a few functions in object type
-    if all_bases and all_bases[-1].__name__ == "pybind11_object":
-        del all_bases[-1]
-    # remove the class itself
-    all_bases = all_bases[1:]
-    # Remove base classes of other bases as redundant.
-    bases: list[type] = []
-    for base in all_bases:
-        if not any(issubclass(b, base) for b in bases):
-            bases.append(base)
-    if bases:
-        bases_str = "(%s)" % ", ".join(
-            strip_or_import(get_type_fullname(base), module, known_modules, imports)
-            for base in bases
-        )
-    else:
-        bases_str = ""
-    if types or static_properties or rw_properties or methods or ro_properties:
-        output.append(f"class {class_name}{bases_str}:")
-        for line in types:
-            if (
-                output
-                and output[-1]
-                and not output[-1].startswith("class")
-                and line.startswith("class")
+        else:  # regular property
+            if readonly:
+                ro_properties.append(f"{self._indent}@property")
+                sig = FunctionSig(name, [ArgSig("self")], inferred_type)
+                ro_properties.append(self._indent + sig.format_sig())
+            else:
+                if inferred_type is None:
+                    inferred_type = self.add_name("_typeshed.Incomplete")
+
+                rw_properties.append(f"{self._indent}{name}: {inferred_type}")
+
+    def get_type_fullname(self, typ: type) -> str:
+        """Given a type, return a string representation"""
+        if typ is Any:
+            return "Any"
+        typename = getattr(typ, "__qualname__", typ.__name__)
+        module_name = self.get_obj_module(typ)
+        assert module_name is not None, typ
+        if module_name != "builtins":
+            typename = f"{module_name}.{typename}"
+        return typename
+
+    def get_base_types(self, obj: type) -> list[str]:
+        all_bases = type.mro(obj)
+        if all_bases[-1] is object:
+            # TODO: Is this always object?
+            del all_bases[-1]
+        # remove pybind11_object. All classes generated by pybind11 have pybind11_object in their MRO,
+        # which only overrides a few functions in object type
+        if all_bases and all_bases[-1].__name__ == "pybind11_object":
+            del all_bases[-1]
+        # remove the class itself
+        all_bases = all_bases[1:]
+        # Remove base classes of other bases as redundant.
```

Comment:
Can't we use `obj.__bases__` instead of `.mro()` so we don't need this?

The incantation to avoid a metaclass that overrides `__bases__` would be `type.__dict__["__bases__"].__get__(obj)`.

```
In [25]: class Meta(type):
    ...:     __bases__ = 42
    ...: 

In [26]: class X(metaclass=Meta):
    ...:     pass
    ...: 

In [27]: X.__bases__
Out[27]: 42

In [29]: type.__dict__["__bases__"].__get__(X)
Out[29]: (object,)
```

Reply:
This code predates my changes.  I'd like to avoid increasing the scope.